src/lib/zkapp.ts

```diff
@@ -611,6 +611,50 @@ class SmartContract {
     party.body.events = Events.pushEvent(party.body.events, eventFields);
   }
 
+  fetchEvents(
+    start: UInt32,
+    end?: UInt32
+  ): { type: string; event: AsFieldElements<any> }[] {
+    // filters all elements so that they are within the given range
+    // only returns { type: "", event: [] } in a flat format
+    let events = Mina.fetchEvents(this.address, this.self.body.tokenId)
+      .filter((el: any) => {
+        let slot = UInt32.from(el.slot);
+        return end === undefined
+          ? start.lte(slot).toBoolean()
+          : start.lte(slot).toBoolean() && slot.lte(end).toBoolean();
+      })
+      .map((el: any) => el.events)
+      .flat();
+
+    // used to match field values back to their original type
+    let sortedEventTypes = Object.keys(this.events).sort();
+
+    return events.map((event: any) => {
```

Comment:
It would probably be smarter to align the events in a stack-like structure, where new events are at the beginning of the array instead of the end. How is it enforced on the protocol level, or rather archive api?

Reply:
hmm, not sure what's better to use 🤔 yeah we should probably mirror whatever the graphql API ends up doing